# Health Data Science CDT - Tutorial: Pathogen sequencing 

## Detecting antimicrobial resistance determinants

### Overview
In this tutorial you will learn how to use various scriptable approaches to detect different antimicrobial resistance determinants. The tutorial will be mostly based using python, but also rely on additional software packages.


### Background
As discussed in the lecture, antimicrobial resistance can be determined by several mechanisms
* Modification of the target
* Degradation of the antimicrobial
* Reduced influx or increased efflux

In turn each mechanism can be brought about by several changes at the genome level
* Single nucleotide polymorphisms (SNPs) in genes or their regulators
* Small insertions or deletions - indels in genes or their regulators
* Gain of new genes on mobile genetic elements including plasmids

Three main apporaches can be used to look for antimicrobial resistance determinants
* Reference based mapping - good for SNPs and some indels, but can't detect novel genetic material like gain of a plasmid
* De novo assembly - good for gain of novel material and larger insertions and deletions

Neither approach works well where there are multiple copies of a gene in a genome - we will develop a bespoke approach for this.


### The world's worst super gonorrhoea
In February 2018, a patient was diagnosed with the most difficult to treat gonorrhoea ever described. You can read more about it here - https://www.bbc.co.uk/news/health-43571120 - and his suscessful treatment here - https://www.bbc.co.uk/news/health-43840505. The more scientific version is here - https://www.eurosurveillance.org/content/10.2807/1560-7917.ES.2018.23.27.1800323 - but I would encourage not to look at this too closely to start with, as it contains many of the answers!

Two isolates from this infection were sequenced using Illumina technology. You are provided with
* Sequence reads - the raw output of the sequencer
* A ready-made mapped concensus fasta file
* A ready-made de novo assembly

The mapped file was generated by mapping reads to a reference genome (e.g. using BWA mem), and searching these for variants from the reference genome (e.g. using samtools mpileup), and quality filtering them to identify variants of high quality. The de novo assembly was made an assembler (e.g. SPAdes).

The infeciton was caused by a gonorrhoea isolate that was resistant to multiple antibiotics - we will attempt to explain these

Antimicrobial|MIC|Interpretation
:--- | :--- | :----
Ceftriaxone	|0.5 mg/L	|Resistant
Cefixime	|2 mg/L	|Resistant
Azithromycin	|> 256 mg/L	|High-level resistant
Ciprofloxacin	|> 32 mg/L	|Resistant
Tetracycline	|32 mg/L	|Resistant
Benzylpenicillin	|1 mg/L	|intermediate susceptible
Spectinomycin	|8 mg/L	|Susceptible
Gentamicin	|2 mg/L	|No resistance breakpoint available (low value)
Ertapenem	|0.032 mg/L	|No resistance breakpoint available (low value)

MIC, minimum inhibitory concentration.

### Suggested resources
You will need to use the following biopython tools in the tutorial
* https://biopython.org/wiki/SeqIO - for reading in the mapped files
* http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc93 - for searching de novo assemblies

You will also need a way to re-map the raw reads to a novel reference
* You can do this with BWA mem - https://github.com/lh3/bwa
* You will need a way to analyse the mapped files - I suggest using pysam to run samtools - https://pysam.readthedocs.io/en/latest/api.html#introduction

Most of the relevant resistance mutations and variants in Neisseria gonorrhoeae (the bacteria that causes gonorrhoea) are catelogued in the NG-STAR database - https://ngstar.canada.ca/alleles/query?lang=en

If you want to know more about the ways in which gonorrhoea can become resistance to commonly used antibiotics this is a good review - https://cmr.asm.org/content/27/3/587.long
<br />
<br />

## Task 1 - Explain ciprofloxacin resistance
**A common mutation causing ciprofloxacin resistance is found in the *gyrA* and results in the following amino acid change - S91F. Can you confirm if this is present in the sequence data, using the mapped data, and find another resistance determinant.**

Tips
* Find the location of the *gyrA* gene in the reference file - https://www.ncbi.nlm.nih.gov/nuccore/NC_011035.1?report=graph - use the find function on the webpage
* Use biopython's `SeqIO.read()` function to read in the mapped fasta file
* You can will need to extract the DNA sequence of the gene using it's coordinates (remember that these number from zero in python, but from one on the NCBI website!)
* You will need to translate the DNA sequences to a protein sequence, `seq.transplate()` in biopython (for *gyrA* the gene is in the same direction as the DNA is numbered, if it were not you would need to generate the reverse complement sequence first), if this works you should end up with a string of amino acids represented as single letters and ending with a stop codon shown as an asterisk.
* Compare the amino acid sequences from the reference with the sequence from the case using python, where are the differenes and what are they?
<br />
<br />

## Task 2 - Explain the tetracycline resistance
**Tetracycline resistance is conferred by gain of a new gene, *tetM*, carried on a plasmid. Is this gene present?**

Tips
* Use the de novo assembly provided
* You will need to create blast database files for the de novo assembly with this command
```makeblastdb -dbtype nucl -in gc.contigs.fa```
* Use the biopython blast module to run a blast command that searches for the tetM gene



# To provide
 - reference genome
 - mapped file
 - assembly
 - raw reads

Read the mapped file 

